README.md
# RealRecombination

Created by Ruth Gómez Graciani

## About

Give project objective, description and scope

## Structure

* `analysis`: a directory for each new run should be created, with the parameter files for that run if necessary.
* `code`: `software/` is for external code and the others are original code.
* `data`: `raw/` for data as-is and `use/` for preprocessed data that is the same regardless of analysis parameters. Please explain origin, content and format of each file in `raw/` to make it reproducible. 
* `project`: papers and other useful bibliography, project history.  Project history should have all pushed tests, even if they were discarded. Discarded tests should be added a comment about why they were discarded. 
* `report`: reports, power points, etc. each in its directory. Final, sent files should be in deliver/ subdirectories.
* `tmp`: temporary files, organized as in `analysis`. 

## Prerequisites

This project should be associated with a conda environment of the same name. To activate it, do:

```conda activate RealRecombination```

## Running the tests

`run_example.sh` runs the analyses with part of the data. 

The analysis is divided in different parts, that are options in `runall.sh`. Each of them could have its own requirements:

* `runall download` downloads available raw data (`data/raw`).
* `runall dbgap` downloads data from dbgap given an ID (Crick).
* `runall preprocess` processes raw data to use it (`data/use`).
* `runall merge` merges test and reference data for all chromosomes in preparation to do the PCA. 
* `runall pca` performs the pca analysis.
* `runall pcaplot` makes the plot for the pca results. It is separated from pca to allow independent adjustments in plots. 
* `runall impute` imputes a list of invesions in a list of individuals with IMPUTE2. 
* `runall imputetables` makes summary tables for the imputation results.
* `runall tagsnps` is to check which tag snps has an inversion given a group of individuals (usually a population).
* `runall crossovers` merges genotype information with crossovers information for the statistical analysis. 

## A note on code comment structure

The header: 
```
#!/bin/bash
# Ruth Gómez Graciani
# 01 04 2020

###############################################################################
# Description:                                                                
#                                         
###############################################################################
```

The levels of code:

```
# MAIN BLOCK
# Description
# =========================================================================== #

# MAIN BLOCK - SUB BLOCK
# --------------------------------------------------------------------------- #

# Normal comment

```

Note the 4-3-2-1 lines in the different comments.



project/history.txt
2020-04-01 RealRecombination structure was created.
2020-04-13 Manually download data for a breakseq test:
	# TOOL VERSION: 2.10.4
	cd data/raw 
	mkdir avery_SRA_key avery_SRA avery_fastq
	# Downloaded SRA key for this dataset prj_21579.ngc into data/raw/avery_SRA_key (manually).
	# Wrote into rawdata readme (manually).
	# Copied directory structure into Crick (manually).
	vdb-config -i
	# I make sure that the "Cache" tab has "local file-caching" enabled and set the "public user-repository" to 20200401_RealRecombination/data/raw/avery_SRA (manually).
	# Prefetch:
	prefetch --max_size 100000000 --ngc avery_SRA_key/prj_21579.ngc  SRR10140422
	prefetch --max_size 100000000 --ngc avery_SRA_key/prj_21579.ngc  SRR10140423
	# Fastq conversion - I store it into Crick:
	cd avery_SRA/sra/sra/
	fastq-dump --ngc ../../../avery_SRA_key/prj_21579.ngc  -O /run/user/1001/gvfs/sftp:host=crick.uab.cat,user=rgomez/home/rgomez/20200401_RealRecombination/data/raw/avery_fastq SRR10140422.sra 
	# Read 440677597 spots for SRR10140422.sra
	# Written 440677597 spots for SRR10140422.sra
	fastq-dump --ngc ../../../avery_SRA_key/prj_21579.ngc  -O ../../../avery_fastq SRR10140423.sra
	# Read 399028996 spots for SRR10140423.sra
	# Written 399028996 spots for SRR10140423.sra
	cd ../../../avery_fastq/
	bgzip SRR10140423.fastq
	mv SRR10140423.fastq.gz /run/user/1001/gvfs/sftp:host=crick.uab.cat,user=rgomez/home/rgomez/20200401_RealRecombination/data/raw/avery_fastq
2020-04-14 Manually copied avery_individuals data.
2020-04-15: runall.sh all -a hg19 -i data/raw/avery_individuals -s 5: DISCARDED because .strand files were wrong.
2020-04-15: runall.sh all -a hg19 -i data/raw/avery_individuals -s 10: Stopped after `download` and `preprocess` to improve `merge`. 
2020-04-16: runall.sh merge -s 5 -r data/raw/1000genomes_hg19 -i data/use/avery_individuals: `merge` successfully finished. Ignored results because they are too heavy for git. 
2020-04-17: Manually copied the results in analysis/2020-04-16_mergePca from Muscul to local.
2020-04-17: Manually copied the file with population information from 1000k genomes to data/raw/1000genomes_hg19/integrated_call_samples_v3.20130502.ALL.panel
2020-04-18: runall.sh pca -f analysis/2020-04-16_mergePca/final_vcf.vcf.gz
2020-04-20: breakseq test worked wonderfully. I'll download the rest of the files. The list is in data/raw/avery_SRA_accessionlist/accession_list.txt but I won't make it automatic due to space issues. I'll put a runall option to do all the downloading and saving into crick.
2020-04-20: runall.sh dbgap -d SRR10140453: Didn't work because license was obsolete 
2020-04-20: runall.sh pcaplot -f analysis/2020-04-18_04_runPca/pcaResult.Rdata -p data/raw/1000genomes_hg19/integrated_call_samples_v3.20130502.ALL.panel
2020-04-21: Manually added previous reports.
2020-04-21: Created or moved parameter and imput files for imputation:
	# Manually copied directory data/raw/1000genomes_hg19_maps
	# Manually copied data/use/avery_individuals/samples_population.txt
	# Changed data/use/avery_individuals/samples_population.txt to make individual 25 AFR
	# Manually copied `inversions_info` into raw and use data.
	mkdir data/use/conditions
	# Manually created imputConditions.txt
2020-04-21: Added IMPUTE2 to the software directory. 
2020-04-21: Added QCTool to the software directory.
2020-04-22: runall.sh impute -f data/use/conditions/inversionToImpute.txt -x data/use/conditions/imputConditions.txt -r data/raw/1000genomes_hg19 -m data/raw/1000genomes_hg19_maps -i data/use/avery_individuals -c data/use/inversions_info/Inversions_coordinates_hg19.csv -g data/use/inversions_info/all_genotypes.csv -s 10: there was a small error in the output files - a tab at the end of each line
2020-04-24: runall.sh impute -f data/use/conditions/inversionToImpute.txt -x data/use/conditions/imputConditions.txt -r data/raw/1000genomes_hg19 -m data/raw/1000genomes_hg19_maps -i data/use/avery_individuals -c data/use/inversions_info/Inversions_coordinates_hg19.csv -g data/use/inversions_info/all_genotypes.csv -s 10: solved the error with the new code but using just the requied bits and maintaining the previous date to avoid re-running all the imputation
2020-04-26: runall.sh imputetables -i analysis/2020-04-22_06_imputation/ -p data/use/avery_individuals/samples_population.txt -t data/use/inversions_info/Inversions_imputability.txt
2020-04-27: Manually created conditions for tag SNP checking
2020-04-27: runall.sh tagsnps -f data/use/conditions/inversionToTag.txt -x data/use/conditions/populationToTag.txt -i data/use/avery_individuals -c data/use/inversions_info/Inversions_coordinates_hg19.csv -g analysis/2020-04-26_07_imputationTables/new_genotypes.csv -s 10: there was a small error in the output file concatenation. DISCARDED to have the same format in all of them.
2020-04-28: runall.sh tagsnps -f data/use/conditions/inversionToTag.txt -x data/use/conditions/populationToTag.txt -i data/use/avery_individuals -c data/use/inversions_info/Inversions_coordinates_hg19.csv -g analysis/2020-04-26_07_imputationTables/new_genotypes.csv -s 10:  solved the error with the new code but using just the requied bits and maintaining the previous date to avoid re-running all the tag SNPs check. DISCARDED to have the same format in all of them. 
2020-04-28: I did a summary genotype file with R:
	R 
			one<-read.table("data/use/inversions_info/all_genotypes.csv", sep = "\t", header = TRUE)
			two<-read.table("analysis/2020-04-26_07_imputationTables/new_genotypes.csv", header = TRUE)
			colnames(two)[1]<-colnames(one)[1]
			genotypes<-plyr::rbind.fill(one, two)
			write.table(genotypes,   "data/use/inversions_info/all+new_genotypes.csv", quote = FALSE, sep = "\t", na = ".", row.names = FALSE, col.names = TRUE) 
2020-04-28: Manually added a population file to analysis/2020-04-16_mergePca to use this dataset 

code/rscript/imputeParse.R
#!/usr/bin/env Rscript
# Ruth Gómez Graciani
# 25 04 2020

###############################################################################
# Description:                                                                
# Make summary tables from IMPUTE2 imputation results                                     
###############################################################################

# LOAD ARGUMENTS 
# =========================================================================== #
args = commandArgs(trailingOnly=TRUE)

# # Test if there is at least one argument: if not, return an error
if (length(args)<4) {
  stop("Not enough input files.", call.=FALSE)
}

# Example
#  
# args[1]<-"analysis/2021-02-16_06_imputation_merged" # Directory with imputation results in .vcf
# args[2]<-"data/use/avery_individuals/samples_population.txt"  # Test individuals population
# args[3]<-"data/use/inversions_info/Inversions_imputability.txt" # Inversions imputability
# args[4]<-"tmp/test" # output directory



# LOAD PACKAGES
# =========================================================================== #

# First specify the packages of interest
packages = c("stringr", "reshape2", "naniar")

# Now load or install&load all
lapply(packages,FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

# BASIC TABLE PARSING
# Results of imputation and extra info are all grouped into a table 
# =========================================================================== #

# BASIC TABLE PARSING - Imputation results
# --------------------------------------------------------------------------- #

# Set it as a function to limit stored intermediate objects
# directory<-args[1]
  parsedir<-function(directory){
    
  # Previous parameters
    files<-list.files(directory, pattern = "*.vcf", full.names=TRUE)
    filelist<-list()
    control_list<-c()
    test_list<-c()

  # Parsing loop
    for (filename in files) {
    trim_filename<-sub( ".*\\/", "", filename)
    trim_filename<-sub( "_readSum.vcf", "", trim_filename)
    table<-read.table( filename, sep = "\t", stringsAsFactors = FALSE, comment.char = "" , skip = 2, header = TRUE, check.names = FALSE )
  
    vcf<-table[,c(3,10:length(table))]
    
    vcf_melted<-reshape2::melt(vcf, id.vars = "ID")
    
    vcf_melted$Inversion<-str_split_fixed(vcf_melted$ID, ";", 2)[,1]
    vcf_melted<-cbind(vcf_melted[,c("Inversion", "variable")], str_split_fixed(vcf_melted$value, ":", 2), stringsAsFactors=FALSE)
    
    # vcf_merged<-merge(vcf_melted, populations , by.x = "variable", by.y = "sample", all.x = TRUE)
    if (grepl("ALL", trim_filename)) {
      defname<-str_split_fixed(trim_filename, "_", 2)[2]
      test_list<-c(test_list, trim_filename)
    }else{
      defname<-"con"
      control_list<-c(control_list, trim_filename)
    }
    colnames(vcf_melted)<-c("Inversion", "Individual", paste( "Genotype", defname, sep="_"), paste( "Probability", defname, sep="_"))
    
    
    filelist[[trim_filename]]<-vcf_melted
    
  }
  
  # Merge controls
    filelist[["control"]]<-do.call("rbind", filelist[control_list])
  
  # Merge all
    all_3c<-Reduce(function(x, y) merge(x, y, by=c("Inversion", "Individual"), all = TRUE), filelist[c(test_list, "control")] )
  
  # Replace "" by NA
    indivs_vector<-all_3c$Individual
    all_3c<- all_3c %>% replace_with_na_all(condition = ~.x == "")
    all_3c<-as.data.frame(all_3c)
    all_3c$Individual<-indivs_vector
    
  # Return table
    return(all_3c)
}

# Apply function
  all_3c<-parsedir(args[1])

# BASIC TABLE PARSING - Add extra info
# --------------------------------------------------------------------------- #
  
# Populations
  populations<-  read.table(args[2], header = TRUE, stringsAsFactors = FALSE)[,c(1,3)]
  colnames(populations)<-c("Individual", "Population")
  
  all_3c<-merge(all_3c, populations, all = TRUE)

# Inversion inputability (sorted factor)
  imputability<-read.table(args[3], header = TRUE)
  imputability$GLB<-factor(imputability$GLB, levels = c("No_Polymorphic", "Tagged", "Imputable", "No_Imputable"))
  imputability$EUR<-factor(imputability$EUR, levels = c("No_Polymorphic", "Tagged", "Imputable", "No_Imputable"))
  imputability$AFR<-factor(imputability$AFR, levels = c("No_Polymorphic", "Tagged", "Imputable", "No_Imputable"))

  all_3c<-merge(all_3c,imputability, all.x = TRUE)

# IMPUTATION SUMMARIZATION
# Results of imputation are classified according to their reliability
# =========================================================================== #
  
# IMPUTATION SUMMARIZATION - Uniformity: are all results equal? 
# DEPRECATED 
# --------------------------------------------------------------------------- #
  
  # all_3c$Uniformity<-""
  # 
  # # Known populations
  # all_3c[which(all_3c$Genotype_100 != all_3c$Genotype_250 | all_3c$Genotype_250!= all_3c$Genotype_500 | all_3c$Genotype_250!= all_3c$Genotype_100),"Uniformity"]<-"bad"
  # all_3c[which(all_3c$Genotype_100 == all_3c$Genotype_con & all_3c$Uniformity == "" ), "Uniformity"]<-"good"
  # all_3c[all_3c$Uniformity == "", "Uniformity"]<-"alert"
  # # Unknown populations
  # all_3c[(all_3c$Genotype_100 != all_3c$Genotype_250 | all_3c$Genotype_250!= all_3c$Genotype_500 | all_3c$Genotype_250!= all_3c$Genotype_100) & all_3c$Population == "ALL","Uniformity"]<-"bad"
  # all_3c[ all_3c$Uniformity != "bad" & all_3c$Population == "ALL", "Uniformity"]<-"good"

# IMPUTATION SUMMARIZATION - Quality: do all results have a probability > 80%?
# DEPRECATED
# --------------------------------------------------------------------------- #  
  
  # all_3c$Quality<-""
  # 
  # #Known populations
  # all_3c[which(all_3c$Probability_100 > 0.8 & all_3c$Probability_250 > 0.8 & all_3c$Probability_500 > 0.8 & all_3c$Probability_con > 0.8 ),"Quality"]<-"good"
  # all_3c[which(all_3c$Probability_100 < 0.8 & all_3c$Probability_250 < 0.8 & all_3c$Probability_500 < 0.8 & all_3c$Probability_con < 0.8 ),"Quality"]<-"bad"
  # all_3c[ all_3c$Quality == "","Quality"]<-"alert"
  # #Unknown populations
  # all_3c[(all_3c$Probability_100 > 0.8 & all_3c$Probability_250 > 0.8 & all_3c$Probability_500 > 0.8 ) & all_3c$Population == "ALL","Quality"]<-"good"
  # all_3c[(all_3c$Probability_100 < 0.8 & all_3c$Probability_250 < 0.8 & all_3c$Probability_500 < 0.8 )& all_3c$Population == "ALL","Quality"]<-"bad"
  # all_3c[ !(all_3c$Quality %in% c("good", "bad")) & all_3c$Population == "ALL","Quality"]<-"alert"

# OUTPUT SPREADSHEET
# Results of imputation are included into a spreadsheet with two sheets
# =========================================================================== #
  
# OUTPUT SPREADSHEET - Sheet 1: Imputation Results
# DEPRECATED
# --------------------------------------------------------------------------- #   

  # sheet1<-all_3c[order(all_3c$GLB, all_3c$Individual),c("Inversion","Individual","Population","Genotype_con","Probability_con","Genotype_100","Probability_100","Genotype_250","Probability_250","Genotype_500","Probability_500","GLB","AFR","EUR","Uniformity","Quality")]  
  # write.csv(sheet1, paste0(args[4],"/imputationResults.csv"), row.names = FALSE)

# OUTPUT SPREADSHEET - Sheet 2: Summary by inversion
# DEPRECATED
# --------------------------------------------------------------------------- #   
# 
#   # Store as function to avoid intermediate objects
#   summarize<-function(all_3c){
#   
#     # List populations
#     my.pops<-c(unique(all_3c$Population))
#   
#     # Start list
#     res_list<-list()
#     
#     for (pop in my.pops) {
#       
#       cols<-10
#       colist<-c("Population", "Imputability","Good_Geno" , "P_pop", "P_100",  "P_250","P_500", "Summary", "Heteros")
#       
#       invs<-unique(all_3c$Inversion)
#       
#       tmp<-data.frame(matrix(vector(), nrow = length(invs), ncol=cols, dimnames = list(c(1:length(invs)), c("Inversion",colist))), stringsAsFactors = FALSE )
#       tmp$Inversion<-invs
#       
#       # For each inversion and population, make a summary
#       for (inv in tmp$Inversion) {
#         
#         tmp_2<-all_3c[all_3c$Inversion == inv & all_3c$Population == pop,]
#         
#         if(pop %in% "ALL"){
#           imput<-as.character(tmp_2[1,"GLB"])
#           popmean<-NA
#         }else if(pop %in% c("EUR", "AFR")){
#           imput<-as.character(tmp_2[1,pop])
#           popmean<- mean(as.numeric(tmp_2$Probability_con))
#         }else{
#           imput<-NA
#           popmean<- mean(as.numeric(tmp_2$Probability_con))
#         }
#         
#         # Content of the summary
#         content<-c(
#           pop,
#           imput,
#           table(tmp_2$Uniformity)["good"]/sum(table(tmp_2$Uniformity)),
#           popmean,
#           mean(as.numeric(tmp_2$Probability_100)),
#           mean(as.numeric(tmp_2$Probability_250)),
#           mean(as.numeric(tmp_2$Probability_500)) ,
#           sum(table(tmp_2[tmp_2$Uniformity == "good","Genotype_100"]) > 0),
#           table(tmp_2[tmp_2$Uniformity == "good","Genotype_100"])["HET"] 
#         )
#         
#         tmp[tmp$Inversion == inv, colist] <- content
#         
#       }
#       
#       res_list[[pop]]<-tmp
#       # rm(tmp, tmp_2, content)
#     }
#     
#     inversion_summary<-do.call(rbind, res_list)
#     inversion_summary<-inversion_summary[order(inversion_summary$Inversion),]
#     
#     inversion_summary[is.na(inversion_summary$Heteros),"Heteros"]<-0
#     
#     return(inversion_summary)
#   }
#   
#   sheet2<-summarize(all_3c)
#   write.csv(sheet2,   paste0(args[4],"/summaryByInversion.csv"), row.names = FALSE) 

# OUTPUT GENOTYPES FILE
# Automatically selected genotypes that can be considered good quality
# =========================================================================== #
  
# OUTPUT GENOTYPES FILE - Definitive accepted genotypes
# --------------------------------------------------------------------------- #  

  # # Take imputable inversions
  # imputable<-as.character(imputability[!(imputability$GLB %in% c("No_Imputable")|imputability$AFR%in% c("No_Imputable")|imputability$EUR%in% c("No_Imputable")), "Inversion"])
  # 
  # # Make it as a function to avoid intermediate variables
  # makeGenotypes<-function(all_3c, imputable){
  #   
  #   # Take interesting inversions
  #   part<-all_3c[all_3c$Inversion %in% imputable, ]
  # 
  #   # Make a case marker
  #   part$accepted<-NA
  # 
  #   # Accepted case 1: good uniformity and good probability
  #   part[part$Uniformity == "good" & 
  #          part$Quality == "good" , "accepted" ]<-"yes"
  #   
  #   # Accepted case 2: good uniformity and alert probability but good probability in population-specific
  #   part[is.na(part$accepted) & part$Uniformity == "good" & part$Quality == "alert" & 
  #          part$Probability_con > 0.8  &  !is.na(part$Probability_con ), "accepted" ]<-"yes"
  #   
  #   # Accepted case 3: FOR LESS STRICT FILTERING - all with good uniformity
  #   part[is.na(part$accepted) & part$Uniformity == "good" , "accepted" ]<-"yes"
  #   
  #   # Accepted case 4: FOR LESS STRICT FILTERING - bad uniformity but good probability in population-specific
  #   part[is.na(part$accepted) & part$Uniformity != "good" & 
  #          part$Probability_con > 0.8  & !is.na(part$Probability_con), "accepted"]<-"pop"
  #   
  #   # Transform to genotypes table
  #   p1<-part[part$accepted %in% c("yes"),c("Inversion", "Individual", "Genotype_100")]
  #   p2<-part[part$accepted %in% c("pop"),c("Inversion", "Individual", "Genotype_con")]
  #   colnames(p1)<-colnames(p2)
  #   p<-rbind(p1,p2 )
  #   w<-reshape2::dcast(p, Individual ~ Inversion)
  #   
  #   return(w)
  # }
  
  # genotypes<-makeGenotypes(all_3c, imputable)
  write.table(all_3c,   paste0(args[4],"/new_genotypes.csv"), quote = FALSE, sep = "\t", na = ".", row.names = FALSE, col.names = TRUE) 
  
  